### PR TITLE
Fix tag-number-pattern and useBranchName configuration options for ContinuousDeployment mode

### PR DESCRIPTION
**Pull request by simonejsing**
Potential fix for issue #902

### CODE REVIEW

File: src/GitVersionCore.Tests/VariableProviderTests.cs [hunk 1]
```diff
@@ -152,4 +152,29 @@ public void ProvidesVariablesInContinuousDeploymentModeForStableWhenCurrentCommi
 
         JsonOutputFormatter.ToJson(vars).ShouldMatchApproved(c => c.SubFolder("Approved"));
     }
+
+    [Test]
+    [Category("NoMono")]
+    [Description("Won't run on Mono due to source information not being available for ShouldMatchApproved.")]
```
Comment (asbjornu) [inline src/GitVersionCore.Tests/VariableProviderTests.cs]: I can't see that `ShouldMatchApproved()` is used in the test, so the `Category` and `Description` attributes can safely be removed, no?

File: src/GitVersionCore.Tests/VariableProviderTests.cs [hunk 2]
```diff
@@ -154,6 +154,8 @@ public void ProvidesVariablesInContinuousDeploymentModeForStableWhenCurrentCommi
     }
 
     [Test]
+    [Category("NoMono")]
+    [Description("Won't run on Mono due to source information not being available for ShouldMatchApproved.")]
```
Comment (asbjornu) [inline src/GitVersionCore.Tests/VariableProviderTests.cs]: [Seems like the build ran on Mono just fine](https://travis-ci.org/GitTools/GitVersion/builds/137612887).

**Comment by asbjornu** [comment]
@simonejsing Since [the build ran on Mono just fine](https://travis-ci.org/GitTools/GitVersion/builds/137612887), I think you can just `reset` back to 228082b8fbcd41062f10c1ff74c9401f487ab85f. Otherwise, this looks good to me. I don't quite understand the need for this, but I see that you and @JakeGinnivan have discussed this in #902, so if he agrees this is a nice-to-have feature, I'll happily merge. :smiley:

File: src/GitVersionCore.Tests/VariableProviderTests.cs (hunk 1 above)
Comment (simonejsing) [inline src/GitVersionCore.Tests/VariableProviderTests.cs]: The first failure I got seemed to be related to the tests running on mono, but you are right it looks like 228082b did pass the build.

**Comment by simonejsing** [comment]
Looks like you were right, the test is enabled again now. Thanks.

**Comment by JakeGinnivan** [comment]
Could we instead change the behavior of `tag-number-pattern` to not replace build metadata but inject itself between the tag name and number.

I don't think we need to support both scenarios.

**Comment by simonejsing** [comment]
@JakeGinnivan you mean for the case where the branch is configured in ContinuousDeployment mode? To clarify, that would essentially mean use the same implementation but just use tag-number-pattern instead of introducing a new option. That would work. The reason I opted for the new construct was to avoid breaking existing behavior, but if that is not a concern then I agree with you.

**Comment by JakeGinnivan** [comment]
`ContinuousDelivery` doesn't really work will pull requests because you can't tag a pull request (it's from another repo) and TBH they can be rebased and such. Meaning we only really have to think about `ContinuousDeployment`.

I would say that the current behaviour is a bug because by it's definition continuous deployment should increment each commit.

**Comment by simonejsing** [comment]
Should we change the default configuration for pull request branches to be ContinuousDeployment as well then? Today it is set to ContinuousDelivery

**Comment by simonejsing** [comment]
I've also noticed that the tag 'useBranchName' does not work when a branch is set to ContinuousDeployment, I'm fixing that as well.

**Comment by simonejsing** [comment]
New review out, let me know if there are more comments.

**Comment by simonejsing** [comment]
Guys can you please review and pull this in?

**Comment by JakeGinnivan** [comment]
Heya

This may have problems with sorting on NuGet. For example PR number 2 will end up higher than pr 11. How will the packages be consumed?

**Comment by simonejsing** [comment]
I see what you mean. Normally it's not a problem since you rarely will consume the latest pre-release packages. I'll add leading zeros like we do for the number.

**Comment by simonejsing** [comment]
Please review

**Comment by asbjornu** [comment]
@simonejsing Looks good, but your branch is a bit of a mess after all the merges. You should instead use rebase, so you don't get all those merge commits. The pull request will look a lot cleaner then. Think you can clean it up a bit? If you need help on how to get it to a cleaner state, please let me know and I'll be happy to let you know exactly what to write on the command line. :smiley:

**Comment by simonejsing** [comment]
Okay, looks like I failed pretty bad at rebasing this...

**Comment by simonejsing** [comment]
That looks better.

**Comment by simonejsing** [comment]
@asbjornu I've rebased onto the latest master and force pushed in order to reuse the PR. Let me know if you would like a new PR instead.

**Comment by asbjornu** [comment]
@simonejsing Looks much better, thanks! 👍

**Pull request merged**